### PR TITLE
Huaweiads - 1. add the corresponding site endpoint selection 2. add package name replacement

### PR DESCRIPTION
**Pull request by wwwyyy**
update huaweiads adapter
1. add the corresponding site endpoint selection 
2. add package name replacement

### CODE REVIEW

**Comment by hhhjort** [comment]
I just want to confirm that you have a reason for routing bid requests to a server based on the location of the of the person viewing the ad rather than the location of the PBS server. The common case we see is putting the endpoint an instance of PBS should use in the PBS config rather than calculating it on the fly as you are doing. Host based config has the advantage of, if the PBS instance is located a long distance from the actual user, you don't then make an equally long call from PBS to the bidder when there is a local bidder available.

For example, if a PBS host has instances set up in both Germany and Singapore, but due to some outage a user in France is routed to the Singapore PBS instance, it may be more efficient to have that PBS instance talk to your asian endpoint rather than your european endpoint.

However if you have your endpoints optimized for local demand, it does make sense to route the requests based on user location rather than PBS host location.

File: adapters/huaweiads/huaweiads.go [hunk 1]
```diff
@@ -278,14 +296,41 @@ func (a *adapter) MakeRequests(openRTBRequest *openrtb2.BidRequest,
...
 	return []*adapters.RequestData{bidRequest}, nil
 }
 
+// countryCode is alpha2, choose the corresponding site end point
+func getFinalEndPoint(countryCode string, defaultEndpoint string) string {
+	if countryCode == "" || len(countryCode) > 2 {
+		return defaultEndpoint
+	}
+
+	var europeanSiteCountryCodeGroup = map[string]string{"AX": "1", "AL": "1", "AD": "1", "AU": "1", "AT": "1", "BE": "1",
```
Comment (hhhjort) [inline adapters/huaweiads/huaweiads.go:313]: It is slightly more efficient to define these as `map[string]interface{}`, as empty interfaces can be the most efficient to store. Also makes it a little more obvious that the value stored in the map has no actual use, it just exists to take advantage of the efficiency of the map's key lookup mechanics.

Comment (SyntaxNode) [reply adapters/huaweiads/huaweiads.go:313]: Is this attempting to simulate a hash set? If so, it's most efficient to use `map[string]struct{}` as the empty struct consumes literally no memory. All the fun idiosyncrasies of Go. See: https://medium.com/@l.peppoloni/how-to-improve-your-go-code-with-empty-structs-3bd0c66bc531

File: adapters/huaweiads/huaweiads.go [hunk 2]
```diff
@@ -337,19 +400,21 @@ func getHeaders(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, request *openrtb2.
...
 // getHuaweiAdsReqJson: get body json for HuaweiAds request
-func getHuaweiAdsReqJson(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest, huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) error {
+func getHuaweiAdsReqJson(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest,
+	huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, extraInfo ExtraInfo) (string, error) {
 	request.Version = huaweiAdxApiVersion
 	var err error
-	if err = getHuaweiAdsReqAppInfo(request, openRTBRequest); err != nil {
-		return err
+	var countryCode string
+	if countryCode, err = getHuaweiAdsReqAppInfo(request, openRTBRequest, extraInfo); err != nil {
```
Comment (hhhjort) [inline adapters/huaweiads/huaweiads.go:408]: you can avoid having to declare `err` and `countryCode` on separate lines by using
`if countryCode, err := getHuaweiAdsReqAppInfo(...`

This is just a stylistic quibble, if you prefer your current construction, that is valid as well.

File: adapters/huaweiads/huaweiads.go [hunk 3]
```diff
@@ -482,9 +547,48 @@ func getHuaweiAdsReqAppInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.
...
+}
+
+// when it has pkgNameConvert (include different rules)
+// 1. when bundleName in ExceptionPkgNames, finalPkgname = bundleName
+// 2. when bundleName conform UnconvertedPkgNames, finalPkgname = ConvertedPkgName
+// 3. when bundleName conform keyword, finalPkgname = ConvertedPkgName
+// 4. when bundleName conform prefix, finalPkgname = ConvertedPkgName
+func getFinalPkgName(bundleName string, extraInfo ExtraInfo) string {
+	for _, convert := range extraInfo.PkgNameConvert {
+		if convert.ConvertedPkgName == "" {
```
Comment (hhhjort) [inline adapters/huaweiads/huaweiads.go:563]: Is there a use case for an `PkgNameConvert` entry to not have a value in `ConvertedPkgName`? If not you may want to check for that when building your adapter and throw an error if an entry is missing this field.

File: adapters/huaweiads/huaweiads.go (hunk 1 above)
Comment (hhhjort) [reply adapters/huaweiads/huaweiads.go:313]: Thanks Scott, I was thinking empty structs but somehow landed on interfaces.

**Comment by mansinahar** [comment]
@wwwyyy Friendly reminder about the above comments whenever you get a chance. Thanks!

**Comment by wwwyyy** [comment]
> I just want to confirm that you have a reason for routing bid requests to a server based on the location of the of the person viewing the ad rather than the location of the PBS server. The common case we see is putting the endpoint an instance of PBS should use in the PBS config rather than calculating it on the fly as you are doing. Host based config has the advantage of, if the PBS instance is located a long distance from the actual user, you don't then make an equally long call from PBS to the bidder when there is a local bidder available.
> 
> For example, if a PBS host has instances set up in both Germany and Singapore, but due to some outage a user in France is routed to the Singapore PBS instance, it may be more efficient to have that PBS instance talk to your asian endpoint rather than your european endpoint.
> 
> However if you have your endpoints optimized for local demand, it does make sense to route the requests based on user location rather than PBS host location.

**Comment by wwwyyy** [comment]
> I just want to confirm that you have a reason for routing bid requests to a server based on the location of the of the person viewing the ad rather than the location of the PBS server. The common case we see is putting the endpoint an instance of PBS should use in the PBS config rather than calculating it on the fly as you are doing. Host based config has the advantage of, if the PBS instance is located a long distance from the actual user, you don't then make an equally long call from PBS to the bidder when there is a local bidder available.
> 
> For example, if a PBS host has instances set up in both Germany and Singapore, but due to some outage a user in France is routed to the Singapore PBS instance, it may be more efficient to have that PBS instance talk to your asian endpoint rather than your european endpoint.
> 
> However if you have your endpoints optimized for local demand, it does make sense to route the requests based on user location rather than PBS host location.

 we are testing with clients, considering privacy concerns, and at the same time，
the data cannot be transferred across sites, and the country will correspond to the determined site，so make changes like this

File: adapters/huaweiads/huaweiads.go (hunk 1 above)
Comment (wwwyyy) [reply adapters/huaweiads/huaweiads.go:313]: Thank you all, it has been revised

File: adapters/huaweiads/huaweiads.go (hunk 2 above)
Comment (wwwyyy) [reply adapters/huaweiads/huaweiads.go:408]: Thank you，done

File: adapters/huaweiads/huaweiads.go (hunk 3 above)
Comment (wwwyyy) [reply adapters/huaweiads/huaweiads.go:563]: thank you, parameter validation has been added, as well as test code

**Comment by wwwyyy** [comment]
> @wwwyyy Friendly reminder about the above comments whenever you get a chance. Thanks!

Thank you, the issues in the code has been corrected, please check, thank you very much for your suggestions

File: adapters/huaweiads/huaweiads.go (hunk 3 above)
Comment (wwwyyy) [reply adapters/huaweiads/huaweiads.go:591]: yes, I know, the check code was added in the Builder() -> getExtraInfo(), getExtraInfo will return the error, Builder()  can catch, you can check.

**Review by hhhjort — APPROVED** [review]
LGTM

**Review by VeronikaSolovei9 — APPROVED** [review]
LGTM

**Pull request merged**